ssh, scp, avahi対応

mnakada opened the pull request:
- 公開鍵をauthorized_keysに書くことでsshでlogin可能
- IP addressでなく[hostnameに書いた名前].localでアクセス可能
- disable_telnetファイルを置くことでtelnetを停止
- 設定ファイルはSD-Cardのtop directoryに置く

rootfs_hack.ext2のbuild環境が違うのでrootfs_hack.ext2は置き換えてもらった方がいいかもしれません
openmiko latest(Release v0.0.27)のbuild時の.configファイルをopenmiko_configとして置いておきます

mnakada committed:
ssh, scp, avahi対応
- 公開鍵をauthorized_keysに書くことでsshでlogin可能
- IP addressでなく[hostnameに書いた名前].localでアクセス可能
- disable_telnetファイルを置くことでtelnetを停止
- 設定ファイルはSD-Cardのtop directoryに置く

rootfs_hack.ext2のbuild環境が違うのでrootfs_hack.ext2は置き換えてもらった方がいいかもしれません
openmiko latest(Release v0.0.27)のbuild時の.configファイルをopenmiko_configとして置いておきます

bakueikozo commented:
お、まさにビルド環境調整しようと思っていたところでした。
今搭載のrootfsはIngenicのgccで作ったやつで、どうにもbuildrootのinternal gccと合わなくて困っていたところです。
open化を考えるとopenmiko環境で完結したほうがいいのでそっちにシフトしていこうと思ってます。

mnakada committed:
openmikoでext2ファイルまで生成するように修正
ext2までbuildできるのに気がついておらず、rootfs.tarからrootfs_hack.ext2を別途生成してました

mnakada commented:
本当はiCamera_appをbuildしているuClibc-0.9.33.2の環境を用意するといいのかもしれませんが、openmikoだとuClibc-ng-1.0.12を使用しているためバージョンの整合性が取れないようです。
Ingenicの環境だとuClibc-ngでなくuClibcの設定があったりしないでしょうか？

bakueikozo commented:
素のIngenicのSDKだとulibc対応のgccとulibcでコンパイルするようになっていますので、
で、ngになってるのはその上の(hlcloudとlocalsdk)ツールキットで対応されてるような気がします。

bakueikozo commented:
ああそうだ、今公開しているrootfsは、IngenicのSDKに入っているgcc4.7.2をExternal toolchainに指定してコンパイルしています。

mnakada commented:
rootfs_hack.ext2の/usr/bin/のv4l2rtspserverやlftpがglibcでbuildされていたのでIngenicのSDKもglibcなのかと考えましたが勘違いしてました。
v4l2rtspserverやlftpはopenmikoでIngenicのgccでbuildしたということですね。
ありがとうございます。

bakueikozo commented:
あーそういうことですか。
Ingenicのgccもulibc、glibcどちらか選べるようにはなってます。
全体で問題が起きなければ、openmikoの(buildroot)のツールチェーンでいいんだとは思います。
カーネルモジュールが上手く入らなかったりしたので各モジュールバラバラに環境作ってくっつけてるので、その辺はこれから整理しなきゃの状態です。

mnakada committed:
/tmp/newroot/以下の実行ファイルを実行時にld.so.1が無くてnot foundとなっていた問題の対応を追加

mnakada commented:
shard libraryの参照をLD_LIBRARY_PATHで切り替えてると、うまく動かない場面が出てくるので検討していました。
openmikoのbuild時に実行ファイルにshared libの検索パスを埋め込めるので、それを使う形に修正します。
atomcam内蔵の実行ファイルは/lib, /usr/libのsoファイルを使用, openmikoでbuildした実行ファイルは/tmp/newroot/lib, /tmp/newroot/usr/libのsoファイルを使用するようにします。
環境をまとめるのでお待ちください。

bakueikozo commented:
自分メモもかねて。
https://suzuki-kengo.net/ld-linux-so-library/
このへんのですかね。
libを統合しようとしてlibpthreadが被ったあたりでちょうどどうしようかと思ってたところで助かります。
インストールor起動時にflashから統合してrootfs再構成しようかと思ってたんですが、これでできるならあとはPATHだけで行けそうですね。

mnakada committed:
atomcam内蔵のコマンドを実行するとsegmentation faultが発生する問題の対応
rootfs_hack.ext2をlinker option -rpath /tmp/newroot/lib -rpath /tmp/newroot/usr/libをつけてbuildするように変更
この影響でubootのbuildでerrorが出るので、不要なubootのbuildを外した

mnakada commented:
基本的には実行ファイルにRPATHが設定されているのでatomcam内蔵のコマンドもopenmikoでbuildしたコマンドもそのまま実行できます。
ただ、scp,ssh,v4l2rtspserverとかのいくつかのコマンドにはRPATHが設定されないようです。
対応として/usr/libに両方のshared libraryを合成するようにしました。
ただし、libgcc_s.so, libgcc_s.so.1, libpthread.so.0が重複していてiCamera_appなどから使われてるので、これらはatomcam側を生かしています。
そのため、libpthreadを使用するv4l2rtspserverはLD_LIBRARY_PATHの指定が必要です。
RPATHが入らないものがあるのが気になりますが、ほとんどのコマンドは大丈夫かと思います。

mnakada commented:
openmikoで生成された/bin, /sbin, /usr/bin, /usr/sbin 以下のファイルを確認しました。
RPATHが設定されてないもの、その中で重複するlibpthread.so.0とlibgcc_s.so.1を参照しているものです。
他はRPATHが設定されていました。
usr/bin/ssh-keygen
usr/bin/ssh-keyscan
usr/bin/ssh
usr/bin/ssh-add
usr/bin/ssh-agent
usr/bin/scp
usr/bin/sftp
usr/sbin/sshd
usr/bin/videocapture
  NEEDED               libpthread.so.0
  NEEDED               libgcc_s.so.1
usr/bin/micropython
  NEEDED               libpthread.so.0
  NEEDED               libgcc_s.so.1
usr/bin/v4l2rtspserver
  NEEDED               libpthread.so.0
  NEEDED               libgcc_s.so.1

mnakada committed:
openmikoで生成した実行ファイルにRUNPATHを設定する

mnakada commented:
独自のbuild設定をしているpackageではBR2_TARGET_LDFLAGSの値が反映されないものがあるようです。
また、ingenic_videocapはCMakeがusr/binにinstallする時にRPATHを消しています。
個別にPackageごとにpatchを当てても他に入れたいコマンドとかが出てくると大変なので、post_build_scriptで/bin, /usr/bin, /sbin, /usr/sbin以下のELFファイルにRUNPATHをまとめて設定するようにしました。

rootfs_fake.ext2のbuild手順です。

1.openmikoのdirectoryでopenmiko_scripts.tgzを展開してください。
openmiko_config
openmiko_post-build.sh
patchelf-all
の３つのファイルができます。
これらはdocker内から呼び出されるのでopenmikoのtop directoryに置いてください。

2.openmikoのdocker環境を立ち上げて入ります。
# docker-compose up -d
# docker-compose exec builder bash
# cd build/buildroot-2016.02/

3.configファイルを置き換えます。
# cp /src/openmiko_config .config
# make oldconfig

4.buildします。
# make clean
# make
makeの処理の中でbuildが完了してext2 imageを作る前にBR2_ROOTFS_POST_BUILD_SCRIPTの設定でopenmiko_post-build.shが呼ばれます。
この中でpatxhelf-all --setを実行し、元々設定されていた/src/buildscripts/post_fakeroot_script.shを呼び出しています。
patchelf-allは実行ファイルをfindで探して対象のものにpatchelfでRUNPATHを追加します。

5.出来上がったext2イメージを外にコピーします。
# cp output/images/rootfs.ext2 /src/rootfs_hack.ext2

mnakada commented:
書き忘れてました。
openmikoはv0.027, v0.028で正常動作を確認しています。

openmikoのHEADはbuildは通りますが、実機でバイナリがSegmentation fault起こします。
kernelのTARBALLの参照先が変わったみたいでライブラリの整合性が取れなくなってるのかもしれません。
v.0.029がでたら確認してみようとおもいます。

mnakada committed:
filesystemのmountの構成を変更しiCamera_appをchroot環境内で動作させることでuClibc環境を分離する

/                    rootfs_hack.ext2
/atom            /mtdblock2 (original rootfs)
/system        /atm/systemのbind
/system/bin  /tmp/system_binのbind
    rmの置き換えとiCamera_app -- IPC --> assis経由でのexec対応パッチを当てるため
    kernelを置き換えるならassisもchroot環境に閉じ込められるのでchroot内のsystem/bin/rmのみ対応必要
/dev, /tmp, /mediaは/側にmountして/atom側はbind

mnakada commented:
kernelを置き換える場合に対応しやすいように、mountの構成を以下のように変更してatomcamの環境を/atom以下にchroot環境として閉じ込めています。RUNPATHの対応は不要なので外しました。
~~~ 
/		rootfs_hack.ext2
/atom		/dev/mtdblock2 (atomcam rootfs)
/system		/dev/mtdblock3 (atomcam system)
/system/bin	/tmp/system_binをbind
/dev, /proc, /sys, /media, /tmpは/にmountして/atom以下にbind
~~~
iCamera_appはコマンドの実行をassis経由で呼び出すようです。これはちょっとはまりました。
assisがpivot_rootの前に起動されていてrootが/のままになっているので、コマンドの呼び出し時のlibrary pathがずれてしまいます。
その対応として、pivot_root後の環境に以下の対応をしています。

- /libraryにld-uClibc*をコピー
- /system/binのELFファイルにLD_LIBRARY_PATHを設定するshell scriptを噛ませる対応

rm.shはassisから呼ばれるのでassis起動時のPATH設定だけにある/system/binに配置することで通常のrmに影響がないように修正しました。
ここはkernelを置き換えてassisもchroot環境に入れた時にlftpにLD_LIBRARY_PATHつける必要があります。

その他
- telnetの設定はhoneylab.cgiの方に移動しました。
- /tmp/mmcは/media/mmcと２重mountで微妙なので/media/mmcに統一しました。
- 各shell scriptを少し整理しました。
- いくつかのlogを/tmp/log/にまとめました。
- openmikoはRelease v0.028を使用しています。

mnakada committed:
make.shでkernelとrootfsを生成しSDCard用のイメージをbuildするように修正
SDCardに
  factory_t31_ZMC6tiIDQN
  rootfs_hack.ext2
  authorized_keys
  hostname
の４つのファイルをコピーすることで起動する

mnakada committed:
NASのCIFS(smb)対応
SD-Cardのsamba対応
下記諸々対応のためwebpageを修正
boa -> lighttpdに変更
port 80に移動、http://atomcam.localでアクセス
- hostname変更対応
- swap sizeは64MB固定
- sambaのon/off対応
- rtspserverのon/off対応
- webhook対応(お試し実装、多分bugあり)
それらに伴いtelnet, ftpは削除
sshのauthorized_keysはweb未対応なのでSD-Cardに書く必要あり

mnakada committed:
typo修正、pathの設定ミス確認もれ修正
README.md追記

mnakada committed:
修正反映イメージ

mnakada committed:
不要ファイルの削除
README.md修正
build.md追加

mnakada committed:
buildコメント修正

mnakada committed:
READMEのpath修正

mnakada committed:
WebUIの build時にnpmのpathが通ってなくてUIが生成されていなかった問題の修正

mnakada committed:
jzmmcのdebug messageが出過ぎてkernelのlogが見えないためdisableにする

mnakada committed:
起動を速くするためinitのatomcam呼び出し順を変更

mnakada committed:
ntpdも起動するよう修正

mnakada committed:
webhookにalarm_video/alarm_pictureの送信を追加

mnakada committed:
Merge branch 'ssh-support' of https://github.com/mnakada/atomcam_tools into ssh-support

* 'ssh-support' of https://github.com/mnakada/atomcam_tools:
  READMEのpath修正

mnakada committed:
release image

mnakada committed:
READMEのリリースイメージlink修正

mnakada committed:
scheduled rebootが設定できていなかった問題の対応
SD-CardをoffにしてもSambaが動いている場合がある問題の対応

mnakada committed:
README修正

mnakada committed:
kernel exception対応
原因：起動後60秒くらいのタイミングでSD-Cardのcd-pin detect interruptが発生して、jzmmc driverが再初期化することがある
そのため、swap,rootfsがr/w不可になりExceptionが発生している
多分、タイミング的にwifiの接続完了くらいなのでmmc1の方のdriverがGPIOの設定等で悪さしてるのかも
暫定対応：初回以降のcd-pin detectを無視する
恒久対応：暇があったら

mnakada committed:
rc scriptの下記問題の修正
- ntpd: 起動に失敗する。DNSが引けるようになる前にntpdを起動してエラーで落ちる
- ntspserver: onにしていない時にエラー表示（表示だけの問題）
- samba: offの時にshutdownでエラー表示（表示だけの問題）
- mmc_detect_test.koが存在していないとエラー表示（表示だけの問題）

mnakada committed:
atomcam_toolsのverもwebUIに表示するよう修正

mnakada committed:
Merge branch 'ssh-support'

* ssh-support:
  atomcam_toolsのverもwebUIに表示するよう修正
  rc scriptの下記問題の修正 - ntpd: 起動に失敗する。DNSが引けるようになる前にntpdを起動してエラーで落ちる - ntspserver: onにしていない時にエラー表示（表示だけの問題） - samba: offの時にshutdownでエラー表示（表示だけの問題） - mmc_detect_test.koが存在していないとエラー表示（表示だけの問題）
  kernel exception対応 原因：起動後60秒くらいのタイミングでSD-Cardのcd-pin detect interruptが発生して、jzmmc driverが再初期化することがある そのため、swap,rootfsがr/w不可になりExceptionが発生している 多分、タイミング的にwifiの接続完了くらいなのでmmc1の方のdriverがGPIOの設定等で悪さしてるのかも 暫定対応：初回以降のcd-pin detectを無視する 恒久対応：暇があったら

mnakada committed:
Update README.md

Releaseのlink修正

mnakada committed:
文言typo修正

mnakada committed:
WebHook通知にdevicenameを追加

mnakada committed:
Merge branch 'ssh-support'

# Conflicts:
#	atomcam_configs/overlay_rootfs/scripts/rm.sh
#	atomcam_configs/overlay_rootfs/scripts/webhook.sh

mnakada committed:
Revert "WebHook通知にdevicenameを追加"

This reverts commit 5126b3eee9486bb2b50fcfddc27d1bbabda8b2b2.

mnakada commented:
ちょっと手を入れ過ぎて申し訳ないので自分のGithubで進めようかと思います。
